lib/gossip/damper.js

```diff
@@ -0,0 +1,542 @@
+    var dampScoresByFlapper = {};
+
+    for (var i = 0; i < allResponses.length; i++) {
+        var dampReqResponse = allResponses[i];
+
+        var responseScores = dampReqResponse.scores;
+        if (!Array.isArray(responseScores)) continue;
+
+        for (var j = 0; j < responseScores.length; j++) {
+            var responseScore = responseScores[j];
+
+            var member = responseScore.member;
+            dampScoresByFlapper[member] = dampScoresByFlapper[member] || [];
+            dampScoresByFlapper[member].push(responseScore);
+        }
+    }
+
+    return dampScoresByFlapper;
+}
+
+function findFlappersToDamp(dampScoresByFlapper, rVal, suppressLimit) {
+    // Evaluate the damp scores collected from the fan-out. For each flapper
+    // find whether all scores exceed the suppress limit. If so, the flapper
+    // should be damped.
+    var membersToDamp = [];
+
+    var flapperAddrs = Object.keys(dampScoresByFlapper);
+    for (var i = 0; i < flapperAddrs.length; i++) {
+        var flapperAddr = flapperAddrs[i];
+
+        var flapperScores = dampScoresByFlapper[flapperAddr];
+        if (flapperScores.length >= rVal &&
+                haveAllScoresExceededLimit(flapperScores)) {
+            membersToDamp.push(flapperAddr);
+        }
+    }
+
+    return membersToDamp;
+
+    function haveAllScoresExceededLimit(scores) {
+        return scores.every(function each(score) {
+            return score.dampScore >= suppressLimit;
+        });
+    }
+}
+
+// This function definition is pulled up and attached to the constructor
+// so that its output can be easily unit tested.
+Damper.getMembersToDamp = function getMembersToDamp(allResponses, rVal, config) {
+    var dampScoresByFlapper = groupDampScoresByFlapper(allResponses);
+    var suppressLimit = config.get('dampScoringSuppressLimit');
+    return findFlappersToDamp(dampScoresByFlapper, rVal, suppressLimit);
+};
+
+Damper.prototype.addDampedMember = function addDampedMember(memberAddr) {
+    this.dampedMembers[memberAddr] = {
+        timestamp: this.Date.now()
+    };
+};
+
+Damper.prototype.addFlapper = function addFlapper(flapper) {
+    if (this.dampedMembers[flapper.address]) {
+        this.logger.debug('ringpop damper already damped member', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    if (this.flappers[flapper.address]) {
+        this.logger.debug('ringpop damper already added flapper', {
+            local: this.ringpop.whoami()
+        });
+        return false;
+    }
+
+    // TODO Limit number of flappers?
+    // TODO Expire flappers?
+
+    this.flappers[flapper.address] = flapper;
+    this.logger.debug('ringpop damper added flapper', {
+        local: this.ringpop.whoami(),
+        flapper: flapper.address
+    });
+    this.ringpop.stat('increment', 'damper.flapper.added');
+    this.ringpop.stat('gauge', 'damper.flappers',
+        Object.keys(this.flappers.length));
```

Comment:
Thanks yea. This stat was added recently. I caught it immediately and forgot to push up the fix.